File: Progetto C++/NegozioBiciclette/CBicicletta.cpp
/*
 * Nome : MTBicicletta.h
 * Autore : Ferrari Lorenzo
 * Descrizione : Implementazione della classe derivata CBicicletta
 */
#include "CBicicletta.h"
#include "MTBicicletta.h"
#include "EBicicletta.h"


#include <iostream>
#include <cstring>
using namespace std;

CBicicletta::CBicicletta(): Bicicletta(),MTBicicletta(),EBicicletta::EBicicletta() {
	 tipoFreno = CBicicletta::tipoFrenoST;
}

CBicicletta::CBicicletta(char *materialeIN, int pesoIN, int dim_cerchioIN, int prezzoIN, int rapportiIN, int velocitaIN,int capBatteriaIN, int tipoFrenoIN) :
			Bicicletta(materialeIN,pesoIN,dim_cerchioIN,prezzoIN),MTBicicletta(rapportiIN),EBicicletta::EBicicletta(velocitaIN,capBatteriaIN){
	tipoFreno = tipoFrenoIN;
}

void CBicicletta::getInfo(){
	cout << "Bicicletta da Corsa:" << endl;
	Bicicletta::getInfo();
	MTBicicletta::printSpecific(); // Risoluzione esplicita del name clashing
	EBicicletta::printSpecific();
	CBicicletta::printSpecific();
}

void CBicicletta::printSpecific(){
	if(CBicicletta::getTipoFreno() == 1) cout << "Tipologia di freno: Architetto";
	else if(CBicicletta::getTipoFreno() == 2) cout << "Tipologia di freno: U-Braker";
	else if(CBicicletta::getTipoFreno() == 3) cout << "Tipologia di freno: V-Braker";
	else cout << "Tipologia di freno: NON SUPPORTATO";
	cout << "\n";
}

string CBicicletta::getTipo(){ return "C";}

int CBicicletta::getTipoFreno(){ return tipoFreno; }

CBicicletta::~CBicicletta() { cout << "Rimozione biciclette da corsa" << endl; }

File: Progetto JAVA/NegozioBicicletteJ/src/negozioBiciclette/CBicicletta.java
/**
 * @author: Lorenzo Ferrari
 * Desctizione: classe derivata dalla classe Bicicletta che mediante interfacce implmenta l' 'ereditarietà multipla' dalle classi MTBicicletta e EBicicletta
 */
package negozioBiciclette;
import java.util.ArrayList;
import java.util.List;

public class CBicicletta extends Bicicletta implements MTBInterface,EBInterface {

	// Campo effettivo della classe derivaata
	private int tipoFreno;
	
	//Campo Statico & Default effettivo della classe 
	private static final int tipoFrenoD = 1;
	
	/** Campi delle due classe MTBicicletta E EBiciclcetta aggiunti in quanto,
		non potendo applicare l'ereditarietà multipla in Java, posso fare 'ereditare' 
		i metodi mediante l'implementazione di interfacce ad hoc, mentre i paramentri delle 
		altre due sottoclassi devono essere ricopiati e creati dalla stessa classe CBicicletta.
	**/
	//Campi delle calssi MTBicicletta e EBicicletta che devono essere istanziati direttamente da CBicicletta
		private int velocita;
		private int capBatteria;
		private int rapporti;
		private List<String> sticker;
		
	//Costruttore di default della classe derivata
	public CBicicletta(){
		super();
		velocita = velocitaD;
		capBatteria = 0;
		rapporti = rapportiD;
		tipoFreno = tipoFrenoD;
	}
	
	//Overload del costruttore
	public CBicicletta(String materiale, double prezzo, int peso, int dim_cerchio, int rapporti, int velocita, int tipoFreno) {
		super(materiale,prezzo,peso,dim_cerchio);
		this.tipoFreno = tipoFreno;
		
		sticker = new ArrayList<String>();
		this.rapporti = rapporti;

		this.velocita = velocita;
		this.capBatteria = 0;
	}
	
	@Override
	public int compareTo(Bicicletta o) {
		if (this.getPrezzo()> o.getPrezzo())
			return 1;
		else if(this.getPrezzo() == o.getPrezzo())
			return 0;
		else
			return -1;
	}

	@Override
	String getTipo() { return "C"; }

	@Override
	String printSpecific() {
		return "Velocita: " + this.velocita + "\n" + "Capacità batteria: " + this.capBatteria + "\n" +
				"Rapporti: " + this.rapporti + "\n" + "Stickers: " + sticker.toString() +"\n" +
				"Tipologia freno: " + this.getTipoFreno() + "\n";
	}

	String getTipoFreno() {
		String result = "";
		switch(this.tipoFreno) {
		case 1: result = "Architetto"; break;
		case 2: result =  "U-brake"; break;
		case 3: result = "V-brake"; break;
		}
		return result;
	}
	
	//toString della classe della derivata
	public String toString() {
		return super.toString() + " " + printSpecific();
	}
	
	// Metodi 'ereditati' dalla classe EBicicletta
	@Override
	public int getVelocita() { return this.velocita; }
	
	@Override
	public int getCapBatteria() { return this.capBatteria; }

	//Metodi 'ereditati' dalla classe MTBicicletta
	@Override
	public int getRapporti() { return this.rapporti; }

	@Override
	public int getNumSticker() { return sticker.size();	}

	@Override
	public void setSticker(String... stickers) {
		for(String s: stickers)
			sticker.add(s);
	}
}
